waku/v2/protocol/waku_rln_relay/rln.nim

```diff
@@ -17,72 +17,100 @@ elif defined(MacOsX):
 ## the return bool value indicates the success or failure of the operation
-## 
+
+proc get_merkle_proof*(ctx: ptr RLN, index: uint, output_buffer: ptr Buffer): bool {.importc: "get_proof".}
+## populates the passed pointer output_buffer with the merkle proof for the leaf at position index in the tree stored by ctx
+## the output_buffer holds a serialized Merkle proof (vector of 32 bytes nodes)
+## the return bool value indicates the success or failure of the operation
+
+proc set_leaf*(ctx: ptr RLN, index: uint, input_buffer: ptr Buffer): bool {.importc: "set_leaf".}
+## sets the leaf at position index in the tree stored by ctx to the value passed by input_buffer
+## the input_buffer holds a serialized leaf of 32 bytes
+## the return bool value indicates the success or failure of the operation
+
+proc set_leaves*(ctx: ptr RLN, input_buffer: ptr Buffer): bool {.importc: "set_leaves".}
+## sets multiple leaves in the tree stored by ctx to the value passed by input_buffer
+## the input_buffer holds a serialized vector of leaves (32 bytes each)
+## leaves are set one after each other starting from index 0
+## the return bool value indicates the success or failure of the operation
+
+proc reset_tree*(ctx: ptr RLN, tree_height: uint): bool {.importc: "set_tree".}
+## resets the tree stored by ctx to the the empty tree (all leaves set to 0) of height tree_height
+## the return bool value indicates the success or failure of the operation
+
 #----------------------------------------------------------------------------------------------
+
 #-------------------------------- zkSNARKs operations -----------------------------------------
-proc key_gen*(ctx: RLN[Bn256], keypair_buffer: ptr Buffer): bool {.importc: "key_gen".}
+proc key_gen*(ctx: ptr RLN, output_buffer: ptr Buffer): bool {.importc: "key_gen".}
 ## generates id key and id commitment key serialized inside keypair_buffer as | id_key <32 bytes>| id_commitment_key <32 bytes> |
 ## id commitment is the poseidon hash of the id key
 ## the return bool value indicates the success or failure of the operation
 
-proc generate_proof*(ctx: RLN[Bn256],
-                    input_buffer: ptr Buffer,
-                    output_buffer: ptr Buffer): bool {.importc: "generate_proof".}
-## input_buffer serialized as  [ id_key<32> | id_index<8> | epoch<32> | signal_len<8> | signal<var> ]
-## output_buffer holds the proof data and should be parsed as |proof<256>|root<32>|epoch<32>|share_x<32>|share_y<32>|nullifier<32>|
+proc generate_proof*(ctx: ptr RLN,
+                         input_buffer: ptr Buffer,
+                         output_buffer: ptr Buffer): bool {.importc: "generate_rln_proof".}
+## input_buffer has to be serialized as [ id_key<32> | id_index<8> | epoch<32> | signal_len<8> | signal<var> ]
+## output_buffer holds the proof data and should be parsed as [ proof<128> | share_y<32> | nullifier<32> | root<32> | epoch<32> | share_x<32> | rln_identifier<32> ]
```

Comment:
How come proof size is different?I find the order of share_/share_y/nullifier a bit odd, any particular reason for changing this order?

Reply:
Proof is 128 bytes since it's compressed: 254 bits for each X coordinate of the 3 proof points + 1 bit for the sign of the y coordinate = 32 bytes per coordinate x 4 coordinates (2x1 over G1, 1x2 over G2) = 128 bytes. Compressed proofs are the default returned by ark-circom create_random_proof_with_reduction and verify_proof, so to make the API the same I should have gone through a decompression and re-compression adding useless overhead. There is probably a way to directly process proofs in uncompressed form (to avoid square roots operation to recover the right y coordinate), but [here](https://github.com/status-im/nwaku/pull/1016#discussion_r905563408) we discussed how we can benchmark if this makes actually sense.

The comment is wrong with respect to the implementation: the current version is 
`[ proof<128> | root<32> | epoch<32> | share_x<32> | share_y<32> | nullifier<32> | rln_identifier<32> ]`

The reason for such order of variables was in order to follow the order of the circuit's variables, cf. https://github.com/privacy-scaling-explorations/rln/blob/616ee9b0b085bdf14e7f39df884496b8e77ddc2f/circuits/rln.circom#L5 and  https://github.com/privacy-scaling-explorations/rln/blob/616ee9b0b085bdf14e7f39df884496b8e77ddc2f/circuits/rln-base.circom#L60-L62. For serialization/deserialization purposes, this order makes more sense to me since this is the order parameters should be passed to the verify_proof function. However, ultimately, I didn't want to change too much nwaku code, so I ultimately decided the keep the previously implemented order and add rln_identifier at the end.
Can be easily changed though, since the zerokit witness structure abstracts the fields order: it is just a matter of redefining the serialization/deserialization order. 

Comment updated in https://github.com/status-im/nwaku/commit/80dddff151187047d98901d13e1e9e2c138f6b2f